packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +46,73 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
+    [handleSignTransactions.pending]: (state) => ({
+        ...state,
+        status: SIGN_STATUS.IN_PROGRESS
+    }),
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
+        const retryTxDirection = error.message.includes('Exceeded the prepaid gas')
+            ? RETRY_TX.INCREASE
+            : error.message.includes('TotalPrepaidGasExceeded')
+                ? RETRY_TX.DECREASE
+                : undefined;
+        
+        const tryRetryTx = retryTxDirection && !state.transactions.some((t) => 
+            t.actions && t.actions.some((a) => a.functionCall && a.functionCall.gas && (
+                (
+                    state.retryTxDirection === RETRY_TX.INCREASE 
+                    && a.functionCall.gas.gt(new BN(RETRY_TX.GAS.MAX))
+                ) || (
+                    state.retryTxDirection === RETRY_TX.DECREASE 
+                    && a.functionCall.gas.lte(new BN(RETRY_TX.GAS.MIN))
+                )
+            ))
+        );
 
-        if (!ready) {
-            return {
-                ...state,
-                status: 'needs-confirmation'
-            };
+        if (retryTxDirection && !tryRetryTx) {
+            Mixpanel.track('SIGN - RETRY - retry limit exceeded');
         }
 
-        if (error) {
-            return {
-                ...state,
-                status: 'error',
-                error: payload
-            };
+        const transactions = cloneDeep(state.transactions);
+        if (tryRetryTx) {
+            transactions.forEach((t, i) => {
+                t.actions && t.actions.forEach((a, j) => {
+                    if(a.functionCall && a.functionCall.gas) {
+                        if ((state.retryTxDirection || retryTxDirection) === RETRY_TX.INCREASE) {
```

Comment:
> This should be clamped at RETRY_TX.GAS.MAX so we never increase an action's allocated gas above the max.

We are checking it in the above `tryRetryTx` section - here we are sure, that we should try to change it.

Reply:
> Why the OR here that uses either `state.retryTxDirection` or the local variable? Seems like this would mean that if we previously got an error that set us up to INCREASE (for example), but then we got an error that doesn't match either of our checks that set `retryTxDirection` to a value other than `undefined`, then this OR would result in us _still increasing gas here_ even though the error we received doesn't indicate that we should either increase or decrease it.

The OR is for the first error,  when we don't have `state.retryTxDirection` set. So we will not do retry with the same values at the first iteration.